doc/compiler/optimizer/OptimizerDesignFeatures.md

```diff
@@ -0,0 +1,303 @@
+- [Speculative optimization and compensation techniques as fallback mechanisms when speculation fails](#speculative-optimization-and-compensation-techniques-as-fallback-mechanisms)
+- [Avoidance of expensive analyses such as Static Single Assignment(SSA), interprocedural analysis, expensive aliasing analysis, etc.](#avoidance-of-expensive-analyses)
+- [IL representation that is more amenable to local analyses](#il-representation)
+- [Cooperative threading model, compiled method meta data in conjunction with a stack walker](#cooperative-threading-model-and-compiled-method-meta-data)
+
+## Heuristics
+
+Methods are the units of compilation in the OpenJ9 JIT compiler.
+
+In the OpenJ9 JIT compiler, compilations are triggered either based on a method having been [invoked a certain number of times](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L525)
+or alternately due to a method getting sampled frequently during the run.
+
+There is a [JIT sampling thread](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L1359)
+that periodically tracks the method at the top of the execution stack on each application thread and keeps
+sampling statistics for each method.
+
+Depending on the runtime in question, the specific thresholds for both invocation counts and sampling counts
+can be configured. Initial compilations are typically driven by a combination of counting and sampling,
+whereas recompilations typically depend on sampling since the overhead of counting can be significant in
+compiled code.
+
+The OpenJ9 JIT compiler supports both [asynchronous](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L764)
+and synchronous models for performing compilations.
+
+In an asynchronous model, the thread that is requesting compilation of a method continues executing the method
+after queueing it for compilation, whereas in a synchronous model, the requesting thread waits until the
+compilation is done.
+
+Most compilations are typically done asynchronously since it is better to allow work to be done on the
+requesting thread, but synchronous compilations are necessary in some cases when it is not feasible to
+execute the method in any other way (more on this later in the section on speculative optimization).
+
+The compiler also supports doing compilations on [multiple threads](https://github.com/eclipse-openj9/openj9/blob/20c2e83228780955fd62e1620160a81e7cbec41e/runtime/compiler/control/CompilationRuntime.hpp#L974)
+concurrently to generate compiled code more quickly.
+
+Performance is not only about peak throughput and other metrics such as start-up time and the time to ramp
+up to peak throughput can be improved significantly by asynchronous compilation and using multiple threads
+to do compilations.
```

Comment:
I'm having trouble parsing this sentence.  Maybe it should read "Performance is not only about peak throughput; other metrics such as. . . ."